src/xc.jl

```diff
@@ -109,7 +109,7 @@ mutable struct XCFuncType
 end
 
 xc_func_alloc() = ccall( (:xc_func_alloc, libxc), Ptr{XCFuncType}, () )
-xc_func_init(p::Ptr{XCFuncType}, functional::Int, nspin::Int) =
+xc_func_init(p::Ptr{XCFuncType}, functional::Integer, nspin::Integer) =
```

Comment:
Why use abstract type here? 
My opinion is to use concrete type in these low-level wrapper functions in `xc.jl`.

Reply:
I would argue completely the opposite way and in fact even suggest to use the abstract types to a much larger extent in the file `src/xc.jl`. (We can discuss details in a separate PR). My reasoning is as follows:

- If you stick with the concrete types, you have to do conversion explicitly. In this case, my code in `Functional.jl` will not work for `Int`, because `libxc` happens to return a different integer type than `Int` in some functions. Notice, that on my machine `Cint` (==`Int32`) is not `Int` (==`Int64`). If, however, you use abstract types, Julia does the conversion for you at the point of the `ccall`. This includes warning if overflows occur to the best of my knowledge. So you save the clutter of doing the conversion on the calling side.
- Generally the philosophy of julia is to stay as abstract as possible and use as little type annotations as required. I think annotating `::Integer` is fair enough, because you are bound to use an integer type to fulfil the `C` interface. `Int`, however, is too specific. Especially for `nspin` theoretically any integer type would do and can be seamlessly converted. In the interest of matching with this philosophy either `Integer` or no type annotation should be used in my opinion.

Reply:
If abstract type is used here, same rules need to applied to all similar functions.
My initial idea when I prepare this project is to use concrete type in these directly wrapper functions, and leave the type things to higher level interface. 
> If, however, you use abstract types, Julia does the conversion for you at the point of the ccall.

I didn't notice that before, if so, your way might be better.

I'll do more research about this, and make a PR to apply this feature to all similar functions.
Would you mind if I add you as collaborator and review my PR?

Reply:
> If abstract type is used here, same rules need to applied to all similar functions.

I would agree with that. 
On that note: I noticed you use `Array{Float64, 1}` in many places. I would suggest to use `Array{Float64}` (kind of a semi-abstract type), such that the user is free to use the array dimension to his liking. The idea is that some users might put different spins in an extra dimension, but some might not ... as long as the memory layout is correct the interface should not care about this.

> Would you mind if I add you as collaborator and review my PR?

Not at all. I am very happy to provide feedback.

Reply:
If abstract type is applied, `Array{Float64}` need also to be modified to `Array{<:Real}` to keep all things abstract?

Reply:
Hmm ... that I am not sure, because this could cause `Array{Float32}` to be upcasted to `Array{Float64}`, which is a much larger memory overhead (imagine the array to contain 10000 elements or so) and probably not what the user expects. I would use `Array{Float64}` and thus force an explicit cast in case the user really wants to pass an `Array{Float32}`.